src/map/map.cpp

```diff
@@ -194,6 +193,312 @@ struct s_generator_options {
+			int16 map_id = map_mapname2mapid(map_name.c_str());
+
+			if (map_id == -1) {
+				this->invalidWarning(mapNode, "Map %s does not exist.\n", map_name.c_str());
+				continue;
+			}
+
+			bool enabled;
+
+			if (!this->asBool(mapNode, map_name, enabled))
+				continue;
+
+			if (enabled) {
+				if (util::vector_exists(zone->maps, map_id)) {
+					this->invalidWarning(mapNode, "Map %s already part of this zone.\n", map_name.c_str());
+					continue;
+				}
+
+				zone->maps.push_back(map_id);
+			} else
+				util::vector_erase_if_exists(zone->maps, map_id);
+		}
+	}
+
+	if (this->nodeExists(node, "Mapflags")) {
+		const auto &mapflagNode = node["Mapflags"];
+
+		// Mapflags are stored as double-key so that duplicate mapflags can be parsed.
+		uint16 mapflag_index = 0;
+
+		for (const auto &it : mapflagNode) {
+			std::string flag_name;
+
+			if (!this->asString(it, "Flag", flag_name))
+				return 0;
+
+			std::string flag_name_constant = "MF_" + flag_name;
+			int64 flag;
+
+			if (!script_get_constant(flag_name_constant.c_str(), &flag)) {
+				this->invalidWarning(mapflagNode, "Mapflag %s does not exist.\n", flag_name.c_str());
+				continue;
+			}
+
+			std::string value;
+
+			if (this->nodeExists(it, "Value")) {
+				if (!this->asString(it, "Value", value))
+					continue;
+			} else
+				value = "1";
+
+			zone->mapflags.insert({ std::pair<int16, uint16>(static_cast<int16>(flag), mapflag_index), value });
+			mapflag_index++;
+		}
+	}
+
+	if (!exists)
+		this->put(zone_id, zone);
+
+	return 1;
+}
+
+/**
+ * Initialize Map Zone data
+ */
+void MapZoneDatabase::loadingFinished() {
+	// Copy Map Zone DB data to the map.
+	// This allows for live modifications to the map without affecting the Map Zone DB.
+	for (const auto &zone : map_zone_db) {
+		for (const auto &map : zone.second->maps) {
+			map_data *mapdata = map_getmapdata(map);
+
+			if (mapdata != nullptr) {
+				mapdata->zone.id = zone.second->id;
+				mapdata->zone.disabled_commands = zone.second->disabled_commands;
+				mapdata->zone.disabled_items = zone.second->disabled_items;
+				mapdata->zone.disabled_skills = zone.second->disabled_skills;
+				mapdata->zone.disabled_statuses = zone.second->disabled_statuses;
+				mapdata->zone.restricted_jobs = zone.second->restricted_jobs;
+
+				// Clear previous mapflags
+				mapdata->initMapFlags();
+				mapdata->skill_damage.clear();
+				mapdata->skill_duration.clear();
+			}
+
+			// Apply mapflags from Map Zone DB
+			for (const auto &flag : zone.second->mapflags) {
+				char flag_name[50] = {}, empty[1] = {};
+
+				if (map_getmapflag_name(static_cast<e_mapflag>(flag.first.first), flag_name))
+					npc_parse_mapflag(const_cast<char *>(map_mapid2mapname(map)), empty, flag_name, const_cast<char *>(flag.second.c_str()), empty, empty, empty);
+				else
+					ShowError("MapZoneDatabase::loadingFinished: Invalid mapflag %s from zone %s.\n", flag_name, script_get_constant_str("MAPTYPE_", zone.first));
+			}
+		}
+	}
```

Comment:
Maybe we could check all zones (of all maps defined in `map_index.txt`) and add a warning when no zones are defined. This could help with map management.

`ShowWarning("MapZoneDatabase::loadingFinished: Map %s has no zone. Defaults to DUMMY_ZONE.\n", ...);`
(`DUMMY_ZONE` would have no restrictions)